### PR TITLE
Split the current "CYTHON_COMPILING_IN_LIMITED_API" macro guard

### PR DESCRIPTION
**Pull request by scoder**
Use separate feature guards for:
1) using a global module state struct ("CYTHON_USE_MODULE_STATE")
2) using PyType_FromSpec() for extension types ("CYTHON_USE_TYPE_SPECS")
3) actual limited-API special casing ("CYTHON_COMPILING_IN_LIMITED_API")

### CODE REVIEW

**Comment by scoder** [comment]
The test failure in `fused_cpdef` is due to #3628.

**Comment by MatzeB** [comment]
Hey, just a heads up: My team has been working in the last weeks on a number of `CYTHON_COMPILING_IN_LIMITED_API` improvements as well. Which we haven't gotten around to push into pull requests yet... (Should probably have contacted you earlier). I will make sure we get this stuff at least uploaded to github somewhere this week.

File: Cython/Utility/ModuleSetupCode.c [hunk 1]
```diff
@@ -564,6 +564,13 @@ class __Pyx_FakeReference {
   #define __Pyx_PyVectorcall_NARGS(n)  (n)
 #endif
 
+// PEP-573
+#if PY_VERSION_HEX >= 0x030900B1
+  #define __PyType_FromModuleAndSpec(m, s, b)  PyType_FromModuleAndSpec(m, s, b)
```
Comment (MatzeB) [inline Cython/Utility/ModuleSetupCode.c]: Just wondering, why isn't this `__Pyx_PyType_FromModuleAndSpec`?

File: Cython/Compiler/Nodes.py [hunk 2]
```diff
@@ -5163,17 +5163,19 @@ def generate_type_ready_code(entry, code, heap_type_bases=False):
...
+                        Naming.module_cname,
                         typeobj_cname,
                         tuple_temp,
                         code.error_goto_if_null(typeobj_cname, entry.pos)))
                 code.put_xdecref_clear(tuple_temp, type=py_object_type)
                 code.funcstate.release_temp(tuple_temp)
             else:
                 code.putln(
-                    "%s = PyType_FromSpec(&%s_spec); %s" % (
+                    "%s = PyType_FromModuleAndSpec(%s, &%s_spec, NULL); %s" % (
```
Comment (MatzeB) [inline Cython/Compiler/Nodes.py]: Should use `__PyType_FromModuleAndSpec`

File: Cython/Utility/ModuleSetupCode.c (hunk 1 above)
Comment (scoder) [reply Cython/Utility/ModuleSetupCode.c]: You seem to be referring to old revisions of the code here.

**Comment by scoder** [comment]
The test failure in `special_methods_T561.pyx` is due to https://bugs.python.org/issue39960 and seems difficult to work around. Probably best to disable the test in CPython versions that do not support it.

**Comment by scoder** [comment]
The fix for `__setattr__` in https://bugs.python.org/issue39960 has been merged into CPython 3.8+.

File: Cython/Utility/ModuleSetupCode.c [hunk 3]
```diff
@@ -704,20 +748,30 @@ static CYTHON_INLINE PyObject * __Pyx_PyDict_GetItemStrWithError(PyObject *dict,
...
 #else
+  #define __Pyx_PyType_GetFlags(tp)   (PyType_GetFlags((PyTypeObject *)tp))
   #define __Pyx_PyType_HasFeature(type, feature)  PyType_HasFeature(type, feature)
   #define __Pyx_PyObject_GetIterNextFunc(obj)  PyIter_Next
 #endif
 
+#if CYTHON_USE_TYPE_SPECS && PY_VERSION_HEX >= 0x03080000
+// In Py3.8+, instances of heap types need to decref their type on deallocation.
+// https://bugs.python.org/issue35810
+#define __Pyx_PyHeapTypeObject_GC_Del(obj)  { \
```
Comment (da-woods) [inline Cython/Utility/ModuleSetupCode.c:765]: Does this mean that you should also include the heap types in tp_traverse and tp_clear? (https://bugs.python.org/issue42972)

Comment (scoder) [reply Cython/Utility/ModuleSetupCode.c:765]: Erm, yes… I hadn't thought about that yet, but sure, it's a live reference that can participate in reference cycles, any time.
Good call.

Comment (scoder) [reply Cython/Utility/ModuleSetupCode.c:765]: I think this is the relevant change: https://github.com/python/cpython/commit/1cf15af9a6f28750f37b08c028ada31d38e818dd
(note that it contains a removal of a previous change that tried a generic implementation).

**Comment by mattip** [comment]
It seems that when using `CYTHON_USE_TYPE_SPECS` fails tests, but the rest is passing. Would it be reasonable to split that out as a separate PR?

**Comment by scoder** [comment]
The only remaining issue with the "type specs" tests is that profiling apparently counts cpdef functions twice, for wrapper and underlying function. That suggests a bug in Cython.

But I agree that we don't have to wait for that. The overall changes are valuable already.

**Comment by scoder** [comment]
There's still https://github.com/cython/cython/pull/3611#discussion_r616883983 to be resolved. Heap types need GC support.

**Comment by scoder** [comment]
Merged. Finally. Let's see how many PRs this breaks. :)

**Pull request merged**